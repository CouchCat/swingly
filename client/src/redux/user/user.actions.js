import { UserActionTypes } from "./user.types";
import {
  createUserProfileDocument,
  getCurrentUser,
  signInWithGithub,
  signInWithGoogle
} from "./user.utils";
import { auth, firestore, getEnvironment } from "../../firebase/firebase.utils";
import { openSnackbar } from "../snackbar/snackbar.actions";
import { fetchUsersArray } from "./user.utils";

const getUserWithProvider = async signInMethod => {
  switch (signInMethod) {
    case "google":
      return await signInWithGoogle();
    case "github":
      return await signInWithGithub();
    default:
      return Error("Invalid provider!");
  }
};

export const setCurrentUser = () => async dispatch => {
  try {
    dispatch({ type: UserActionTypes.USER_LOADING });

    const userAuth = await getCurrentUser();
    if (!userAuth) {
      dispatch({ type: UserActionTypes.USER_NOT_LOADING });
      return;
    }

    const userRef = await createUserProfileDocument(userAuth);
    const userSnapshot = await userRef.get();

    dispatch({
      type: UserActionTypes.SET_CURRENT_USER_SUCCESS,
      payload: userSnapshot.data()
    });
  } catch (error) {
    dispatch({
      type: UserActionTypes.SET_CURRENT_USER_FAILED,
      payload: error
    });
  }
};

export const setUsers = () => async (dispatch) => {
  try {
    dispatch({ type: UserActionTypes.USER_LOADING });

    const usersObj = await fetchUsersArray();

    if (Object.keys(usersObj).length) {
      dispatch({
        type: UserActionTypes.SET_USERS_SUCCESS,
        payload: usersObj
      });
    } else {
      dispatch({
        type: UserActionTypes.SET_USERS_FAILED,
        payload: 'No users in DB!'
      });
    }
  } catch (error) {
    dispatch({
      type: UserActionTypes.SET_USERS_FAILED,
      payload: 'Fetch Users failed!'
    });
    dispatch(openSnackbar("error", "Fetch Users failed!"));
  }
}

export const signInWithProvider = signInMethod => async dispatch => {
  try {
    dispatch({ type: UserActionTypes.USER_LOADING });

    const { user } = await getUserWithProvider(signInMethod);
    const userRef = await createUserProfileDocument(user);
    const userSnapshot = await userRef.get();

    dispatch({
      type: UserActionTypes.SIGNIN_SUCCESS,
      payload: userSnapshot.data()
    });

    dispatch(openSnackbar("success", "You have been successfully signed in!"));
  } catch (error) {
    dispatch({
      type: UserActionTypes.SIGNIN_FAILED,
      payload: error
    });

    dispatch(openSnackbar("error", "Sign in failed!"));
  }
};

export const signOut = () => async dispatch => {
  try {
    await auth.signOut();
    dispatch({ type: UserActionTypes.SIGNOUT_SUCCESS });

    dispatch(openSnackbar("success", "You have been successfully signed out!"));
  } catch (error) {
    dispatch({
      type: UserActionTypes.SIGNOUT_FAILED,
      payload: error
    });

    dispatch(openSnackbar("error", "Sign out failed!"));
  }
};

export const signUp = values => async dispatch => {
  const usersSnap = await firestore.collection(`${getEnvironment()}/data/users`).get();
  let userExists = null;
  if (usersSnap) {
    usersSnap.forEach(doc => {
      if (doc.data().email === values.email) {
        userExists = doc.data();
      }
    });
  }

  if (!userExists) {
    try {
      const { user } = await auth.createUserWithEmailAndPassword(
        values.email,
        values.password
      );

      if (user) {
        const userAuth = await getCurrentUser();
        const userRef = await createUserProfileDocument(userAuth, {
          username: values.username
        });

        const userSnapshot = await userRef.get();
        dispatch({
          type: UserActionTypes.SIGNUP_SUCCESS,
          payload: userSnapshot.data()
        });

        await auth.signOut();

        dispatch(openSnackbar("success", "Sign up success!"));
      }
    } catch (error) {
      dispatch({
        type: UserActionTypes.SIGNUP_FAILED,
        payload: error
      });

      dispatch(openSnackbar("error", "Sign up failed!"));
    }
  } else {
    dispatch({ type: UserActionTypes.SIGNUP_FAILED });
    dispatch(openSnackbar("error", "User with given email already exists!"));
  }
};

export const signInWithEmail = values => async dispatch => {
  try {
    const { user } = await auth.signInWithEmailAndPassword(
      values.email,
      values.password
    );

    const userRef = await createUserProfileDocument(user);
    const userSnap = await userRef.get();

    dispatch({
      type: UserActionTypes.SIGNIN_SUCCESS,
      payload: userSnap.data()
    });

    dispatch(openSnackbar("success", "You have been successfully signed in!"));
  } catch (error) {
    dispatch({
      type: UserActionTypes.SIGNIN_FAILED,
      payload: error
    });

    dispatch(
      openSnackbar("error", "Incorrect email or password! Please try again.")
    );
  }
};

export const editUser = (values, currentUser) => async dispatch => {
  let newValues = {};
  if (values.username !== currentUser.username) {
    newValues.username = values.username;
  }

  if (values.description !== currentUser.description) {
    newValues.description = values.description;
  }

  if (Object.keys(newValues).length !== 0) {
    const userRef = firestore.doc(`${getEnvironment()}/data/users/${currentUser.id}`);
    const userSnap = await userRef.get();

    if (userSnap.exists) {
      try {
        const batch = firestore.batch();
        Object.entries(newValues).forEach(([key, val]) => {
          batch.update(userRef, {
            [key]: val
          });
        });
        await batch.commit();

        dispatch({
          type: UserActionTypes.EDIT_SUCCESS,
          payload: newValues
        });

        dispatch(openSnackbar("success", "Edit successful!"));
      } catch (error) {
        dispatch({
          type: UserActionTypes.EDIT_FAILED,
          payload: error
        });
        dispatch(
          openSnackbar("error", "Edit failed! User snapshot does not exist!")
        );
      }
    }
  } else {
    dispatch(openSnackbar("error", "Edit failed!"));
  }
};

export const deleteUser = () => async dispatch => {
  const user = await auth.currentUser;
  
  if (user) {
    try {
      await firestore.doc(`${getEnvironment()}/data/users/${user.uid}`).delete();
      await user.delete();
      dispatch({ type: UserActionTypes.DELETE_SUCCESS });
      dispatch(openSnackbar("success", "Delete successful!"));
    } catch (error) {
      dispatch({ 
        type: UserActionTypes.DELETE_FAILED,
        payload: error
      });
      dispatch(openSnackbar("error", "Deletion failed!"));
    }
  } else {
    dispatch({ type: UserActionTypes.DELETE_FAILED });    
    dispatch(openSnackbar("error", "Deletion failed! User not signed in"));
  }
};
