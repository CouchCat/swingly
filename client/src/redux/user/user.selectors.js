import { createSelector } from 'reselect';
import memoize from 'lodash/memoize';

const selectUser = state => state.user;

export const selectCurrentUser = createSelector(
  [selectUser],
  (user) => user.currentUser
);

export const selectUserIsLoading = createSelector(
  [selectUser],
  (user) => user.isLoading
);

export const selectUsers = createSelector(
  [selectUser],
  (user) => user.users
);

export const selectUserById = createSelector(
  [selectUsers],
  users => memoize(userId => {
    return users.hasOwnProperty(userId) ? users[userId] : false;
  })
);

