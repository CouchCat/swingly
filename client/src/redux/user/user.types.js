export const UserActionTypes = {
  USER_LOADING: 'USER_LOADING',
  USER_NOT_LOADING: 'USER_NOT_LOADING', 
  SET_CURRENT_USER_SUCCESS: 'SET_CURRENT_USER_SUCCESS',
  SET_CURRENT_USER_FAILED: 'SET_CURRENT_USER_FAILED',
  SET_USERS_SUCCESS: 'SET_USERS_SUCCESS',
  SET_USERS_FAILED: 'SET_USERS_FAILED',
  SIGNUP_SUCCESS: 'SIGNUP_SUCCESS',
  SIGNUP_FAILED: 'SIGNUP_FAILED',
  SIGNIN_SUCCESS: 'SIGNIN_SUCCESS',
  SIGNIN_FAILED: 'SIGNIN_FAILED',
  SIGNOUT_SUCCESS: 'SIGNOUT_SUCCESS',
  SIGNOUT_FAILED: 'SIGNOUT_FAILED',
  EDIT_SUCCESS: 'EDIT_SUCCESS',
  EDIT_FAILED: 'EDIT_FAILED',
  DELETE_SUCCESS: 'DELETE_SUCCESS',
  DELETE_FAILED: 'DELETE_FAILED',
  FOLLOW_SUCCESS: 'FOLLOW_SUCCESS',
  FOLLOW_FAILED: 'FOLLOW_FAILED',
  UNFOLLOW_FAILED: 'UNFOLLOW_FAILED',
  UNFOLLOW_SUCCESS: 'UNFOLLOW_SUCCESS',
  TOGGLE_FOLLOWED_USER_SUCCESS: 'TOGGLE_FOLLOWED_USER_SUCCESS',
  TOGGLE_FOLLOWED_USER_FAILED: 'TOGGLE_FOLLOWED_USER_FAILED',
};