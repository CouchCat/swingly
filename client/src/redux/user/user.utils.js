import { firestore, getRootPath } from "../../firebase/firebase.utils"


export const fetchUsersArray = async () => {
  const usersSnap = await firestore.collection(`${getRootPath()}/data/users`).get();

  if (usersSnap.docs.length) {
    return arrayToObject(usersSnap.docs);
  } else {
    throw new Error('No Users found!');
  }
}

export const arrayToObject = (array) => {
  return array.reduce((prev, current) => {
    prev[current.data().id] = current.data();
    return prev;
  }, {});
}