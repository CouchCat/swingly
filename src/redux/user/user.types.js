export const UserActionTypes = {
  SET_CURRENT_USER_SUCCESS: 'SET_CURRENT_USER_SUCCESS',
  SET_CURRENT_USER_FAILED: 'SET_CURRENT_USER_FAILED',
  SIGNIN_SUCCESS: 'SIGNIN_SUCCESS',
  SIGNIN_FAILED: 'SIGNIN_FAILED',
  SIGNOUT_SUCCESS: 'SIGNOUT_SUCCESS',
  SIGNOUT_FAILED: 'SIGNOUT_FAILED',
};