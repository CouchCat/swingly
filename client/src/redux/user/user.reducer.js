import { UserActionTypes } from './user.types';

const INITIAL_STATE = {
  currentUser: null,
  users: {},
  error: null,
  isLoading: false,
};

const userReducer = (state = INITIAL_STATE, action) => {
  switch(action.type) {
    case UserActionTypes.USER_LOADING:
      return {
        ...state,
        isLoading: true
      }
    case UserActionTypes.USER_NOT_LOADING:
      return {
        ...state,
        isLoading: false
      };
    case UserActionTypes.SET_CURRENT_USER_SUCCESS:
    case UserActionTypes.SIGNIN_SUCCESS:
      return {
        ...state,
        currentUser: action.payload,
        error: null,
        isLoading: false
      };
    case UserActionTypes.SET_USERS_SUCCESS:
      return {
        ...state,
        isLoading: false,
        error: null,
        users: action.payload
      }
    case UserActionTypes.SIGNUP_SUCCESS:
      return {
        ...state,
        currentUser: null,
        users: {
          ...state.users,
          [action.payload.id]: action.payload
        }
      }
    case UserActionTypes.SIGNOUT_SUCCESS: 
      return {
        ...state,
        currentUser: null
      }
    case UserActionTypes.DELETE_SUCCESS:
      const newUsers = { ...state.users };
      delete newUsers[state.currentUser];

      return {
        ...state,
        currentUser: null,
        users: newUsers,
        error: null
      };
    case UserActionTypes.EDIT_SUCCESS:
      return {
        ...state,
        users: {
          ...state.users,
          [state.currentUser]: {
            ...state.users[state.currentUser],
            ...action.payload
          }
        }
      };
    case UserActionTypes.FOLLOW_SUCCESS:
      return {
        ...state,
        users: {
          ...state.users,
          [state.currentUser]: {
            ...state.users[state.currentUser],
            following: {
              ...state.users[state.currentUser].following,
              [action.payload]: true
            }
          }
        }
      }
    case UserActionTypes.UNFOLLOW_SUCCESS:
      const newFollowing = { ...state.users[state.currentUser].following };
      delete newFollowing[action.payload];

      return {
        ...state,
        users: {
          ...state.users,
          [state.currentUser]: {
            ...state.users[state.currentUser],
            following: newFollowing
          }
        }
      }
    case UserActionTypes.SIGNUP_FAILED:
    case UserActionTypes.SET_CURRENT_USER_FAILED:
    case UserActionTypes.SET_USERS_FAILED:
    case UserActionTypes.SIGNIN_FAILED:
    case UserActionTypes.SIGNOUT_FAILED:
    case UserActionTypes.EDIT_FAILED:
    case UserActionTypes.DELETE_FAILED:
    case UserActionTypes.FOLLOW_FAILED:
    case UserActionTypes.UNFOLLOW_FAILED:
      return {
        ...state,
        error: action.payload,
        isLoading: false
      };
    default:
      return state;
  }
};

export default userReducer;