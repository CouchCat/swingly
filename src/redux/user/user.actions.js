import { UserActionTypes } from './user.types';
import { auth, getCurrentUser, createUserProfileDocument } from '../../firebase/firebase.utils';

// export const setCurrentUser = user => ({
//   type: UserActionTypes.SET_CURRENT_USER,
//   payload: user
// });

export const setCurrentUser = () => async dispatch => {
  try {
    const userAuth = await getCurrentUser();
    if (!userAuth) return;

    const userRef = await createUserProfileDocument(userAuth);
    const userSnapshot = await userRef.get();
    dispatch({
      type: UserActionTypes.SET_CURRENT_USER_SUCCESS,
      payload: userSnapshot.data()
    });
  } catch (error) {
    dispatch({
      type: UserActionTypes.SET_CURRENT_USER_FAILED,
      payload: error
    });
  }
}
