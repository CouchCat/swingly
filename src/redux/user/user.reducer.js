import { UserActionTypes } from './user.types';

const INITIAL_STATE = {
  currentUser: null,
  error: null,
  loading: false
};

const userReducer = (state = INITIAL_STATE, action) => {
  switch(action.type) {
    case UserActionTypes.USER_LOADING:
      return {
        ...state,
        loading: true
      }
    case UserActionTypes.SET_CURRENT_USER_SUCCESS:
    case UserActionTypes.SIGNIN_SUCCESS:
      return {
        ...state,
        currentUser: action.payload,
        error: null,
        loading: false
      };
    case UserActionTypes.SIGNOUT_SUCCESS:
      return {
        ...state,
        currentUser: null,
        error: null
      };
    case UserActionTypes.SET_CURRENT_USER_FAILED:
    case UserActionTypes.SIGNIN_FAILED:
    case UserActionTypes.SIGNOUT_FAILED:
      return {
        ...state,
        error: action.payload,
        loading: false
      };
    default:
      return state;
  }
};

export default userReducer;