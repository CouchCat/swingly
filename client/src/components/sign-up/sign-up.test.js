import React from "react";
import { render, fireEvent } from "../../test-utils";
import ConnectedSignUp, { SignUp } from "./sign-up.component";
import * as userActionsMock from "../../redux/user/user.actions";

jest.mock("../../redux/user/user.actions");

describe("Sign up component", () => {
  let isOpenMock = true;
  let setOpenMock = jest.fn(); 
  let openSignInMock = jest.fn(); 
  let signUpMock = jest.fn();
  let mockProps = null;

  const applyMockProps = () => {
    mockProps = {
      isOpen: isOpenMock,
      setOpen: setOpenMock,
      openSignIn: openSignInMock,
      signUp: signUpMock
    }
  }

  afterEach(() => {
    isOpenMock = true;
    jest.clearAllMocks();
  });

  describe("Rendering the component", () => {
    it("should render", () => {
      applyMockProps();
      const renderResult = render(<SignUp {...mockProps} />);
  
      expect(renderResult).toMatchSnapshot();
    });
  
    it("should not display the sign-up component when isOpen is false", () => {
      isOpenMock = false;
      applyMockProps();
      const { queryByTestId } = render(<SignUp {...mockProps} />);
  
      expect(queryByTestId("sign-up")).toBeNull();
    });

    it("should display the sign-up component when isOpen is true", () => {
      applyMockProps();
      const { queryByTestId } = render(<SignUp {...mockProps} />);
  
      expect(queryByTestId("sign-up")).toBeTruthy();
    });
  });

  describe("submitting the form", () => {
    it("should signup action should not throw on submit", () => {
      applyMockProps();
      const { getByLabelText, getByText } = render(<ConnectedSignUp {...mockProps} />);

      fireEvent.change(getByLabelText("Username"), { target: { value: "test_user" } });
      fireEvent.blur(getByLabelText("Username"));
      fireEvent.change(getByLabelText("Email"), { target: { value: "test_email@test.com" } });
      fireEvent.blur(getByLabelText("Email"));
      fireEvent.change(getByLabelText("Password"), { target: { value: "12345678" } });
      fireEvent.blur(getByLabelText("Password"));
      fireEvent.change(getByLabelText("Confirm Password"), { target: { value: "12345678" } });
      fireEvent.blur(getByLabelText("Confirm Password"));      
      fireEvent.click(getByText("Submit"));

      expect(userActionsMock.signUp).toHaveBeenCalledTimes(1);
    });

    // Validations are not working correctly (during tests only)
    xit("should keep submit button disabled if values have validation errors", () => {
      applyMockProps();
      const { getByLabelText, getByText } = render(<ConnectedSignUp {...mockProps} />);

      fireEvent.blur(getByLabelText("Username"));
      fireEvent.blur(getByLabelText("Email"));
      fireEvent.blur(getByLabelText("Password"));
      fireEvent.change(getByLabelText("Confirm Password"), { target: { value: "12345678" } });
      fireEvent.blur(getByLabelText("Confirm Password"));      

      fireEvent.click(getByText("Submit"));

      expect(userActionsMock.signUp).toHaveBeenCalledTimes(0);
      expect(getByText("Submit").getAttribute("disabled")).toBe("true");
    });
  });
});