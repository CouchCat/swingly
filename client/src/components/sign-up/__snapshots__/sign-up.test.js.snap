// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign up component Rendering the component should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-VigVT hoGEhl"
        direction=""
      >
        <div
          class="sc-jTzLTM hlpuqw"
        >
          Register
        </div>
        <div
          class="sc-htpNat IDNFI"
          data-testid="sign-up"
        >
          <div
            class="sc-EHOje edRFCv"
          >
            <form
              data-testid="sign-up-form"
            >
              <div
                class="sc-htoDjs jwAKoT"
              >
                <input
                  class="sc-iwsKbI ffGOFQ"
                  id="username"
                  label="Username"
                  name="username"
                  value=""
                />
                <label
                  class="sc-dnqmqq dIwYQT"
                  for="username"
                >
                  Username
                </label>
                <div
                  class="sc-gqjmRU dNgXmx"
                />
              </div>
              <div
                class="sc-htoDjs jwAKoT"
              >
                <input
                  class="sc-iwsKbI ffGOFQ"
                  id="email"
                  label="Email"
                  name="email"
                  value=""
                />
                <label
                  class="sc-dnqmqq dIwYQT"
                  for="email"
                >
                  Email
                </label>
                <div
                  class="sc-gqjmRU dNgXmx"
                />
              </div>
              <div
                class="sc-htoDjs jwAKoT"
              >
                <input
                  class="sc-iwsKbI irafpo"
                  id="password"
                  label="Password"
                  name="password"
                  type="password"
                  value=""
                />
                <label
                  class="sc-dnqmqq dIwYQT"
                  for="password"
                >
                  Password
                </label>
                <div
                  class="sc-gqjmRU dNgXmx"
                />
              </div>
              <div
                class="sc-htoDjs jwAKoT"
              >
                <input
                  class="sc-iwsKbI irafpo"
                  id="confirmPassword"
                  label="Confirm Password"
                  name="confirmPassword"
                  type="password"
                  value=""
                />
                <label
                  class="sc-dnqmqq dIwYQT"
                  for="confirmPassword"
                >
                  Confirm Password
                </label>
                <div
                  class="sc-gqjmRU dNgXmx"
                />
              </div>
              <div
                class="sc-bZQynM haTaOE"
              >
                <button
                  class="sc-bdVaJa ickuCg sc-bwzfXH dEZBwm sc-gzVnrw kWMqOl"
                  data-testid="submit-button"
                  disabled=""
                  type="button"
                >
                  Submit
                </button>
                <button
                  class="sc-bdVaJa jsLivE sc-bwzfXH dEZBwm sc-gzVnrw kWMqOl"
                  type="button"
                >
                  Reset
                </button>
              </div>
            </form>
          </div>
          <div
            class="sc-ifAKCX kgfWhF"
          >
            <button
              class="sc-bdVaJa jsLivE sc-bwzfXH dEZBwm sc-bxivhb lcNDmV"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-left fa-w-10 fa-null fa-rotate-null fa-pull-null "
                data-icon="chevron-left"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
                  fill="currentColor"
                />
              </svg>
              Back
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-VigVT hoGEhl"
      direction=""
    >
      <div
        class="sc-jTzLTM hlpuqw"
      >
        Register
      </div>
      <div
        class="sc-htpNat IDNFI"
        data-testid="sign-up"
      >
        <div
          class="sc-EHOje edRFCv"
        >
          <form
            data-testid="sign-up-form"
          >
            <div
              class="sc-htoDjs jwAKoT"
            >
              <input
                class="sc-iwsKbI ffGOFQ"
                id="username"
                label="Username"
                name="username"
                value=""
              />
              <label
                class="sc-dnqmqq dIwYQT"
                for="username"
              >
                Username
              </label>
              <div
                class="sc-gqjmRU dNgXmx"
              />
            </div>
            <div
              class="sc-htoDjs jwAKoT"
            >
              <input
                class="sc-iwsKbI ffGOFQ"
                id="email"
                label="Email"
                name="email"
                value=""
              />
              <label
                class="sc-dnqmqq dIwYQT"
                for="email"
              >
                Email
              </label>
              <div
                class="sc-gqjmRU dNgXmx"
              />
            </div>
            <div
              class="sc-htoDjs jwAKoT"
            >
              <input
                class="sc-iwsKbI irafpo"
                id="password"
                label="Password"
                name="password"
                type="password"
                value=""
              />
              <label
                class="sc-dnqmqq dIwYQT"
                for="password"
              >
                Password
              </label>
              <div
                class="sc-gqjmRU dNgXmx"
              />
            </div>
            <div
              class="sc-htoDjs jwAKoT"
            >
              <input
                class="sc-iwsKbI irafpo"
                id="confirmPassword"
                label="Confirm Password"
                name="confirmPassword"
                type="password"
                value=""
              />
              <label
                class="sc-dnqmqq dIwYQT"
                for="confirmPassword"
              >
                Confirm Password
              </label>
              <div
                class="sc-gqjmRU dNgXmx"
              />
            </div>
            <div
              class="sc-bZQynM haTaOE"
            >
              <button
                class="sc-bdVaJa ickuCg sc-bwzfXH dEZBwm sc-gzVnrw kWMqOl"
                data-testid="submit-button"
                disabled=""
                type="button"
              >
                Submit
              </button>
              <button
                class="sc-bdVaJa jsLivE sc-bwzfXH dEZBwm sc-gzVnrw kWMqOl"
                type="button"
              >
                Reset
              </button>
            </div>
          </form>
        </div>
        <div
          class="sc-ifAKCX kgfWhF"
        >
          <button
            class="sc-bdVaJa jsLivE sc-bwzfXH dEZBwm sc-bxivhb lcNDmV"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-left fa-w-10 fa-null fa-rotate-null fa-pull-null "
              data-icon="chevron-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
                fill="currentColor"
              />
            </svg>
            Back
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
